Go HTTP-only, part 1

trustin opened the pull request:
Motivation:

Providing a protocol-agnostic API is a noble goal, but in practice it's
hard to balance between API user experience and generalization. I'd like
to make our server-side API go HTTP-only so that we can improve the
overall user experience.

Modifications:

- Add the notion of path parameters (a.k.a path variables) to
  ServiceRequestContext
  - Make PathMapping.apply() return a PathMappingResult
  - Make PathMapping.apply() require a query string so that PathMapping
    can match a query string or replace a query string for internal
    redirection
  - Replace the notion of 'mapped path' with 'path without prefix'
  - Rename PathParamExtractor into DefaultPathMapping
  - Allow a user specify the path pattern with path parameters when
    binding a service with serviceAt()
  - Support path parameters for RegexPathMapping
- Make sure that the request/response type of a service is always
  HttpRequest and HttpResponse by changing the type signature of service
  binding methods from `Service<?, ?>` to `Service<? super HttpRequest, ? extends HttpResponse>`
  - Add LoggingService.newDecorator() because method reference does not
    work anymore because of the type parameter change
- Split RequestContext.path() into path() and query()
- Make RequestContext.fragment() nullable to distinguish between
  non-existent fragment and empty fragment
- Change the type of RequestContext.method() to HttpMethod
- Miscellaneous:
  - Add LoggingClient.newDecorator() for consistency
  - Make sure RequestContext.path() is always raw (undecoded)
  - Extract path and query validation and sanitization logic into
    ArmeriaHttpUtil

Result:

- A user can use path parameters without using DynamicHttpService.
- ServerBuilder expects a user specifies a `Service<HttpRequest, HttpResponse>`
  explicitly. This means more type safety in general.

To-dos:

- Add the path overlap detection back again
- Merge DynamicHttpService into the core construction logic
- Overall package reorganization; there's no point of having 'http in
  a package name once we go HTTP-only.

trustin commented:
Please skip reviewing the classes related with `DynamicHttpService` until I add them back. Meanwhile, please focus on the public API changes.

trustin commented on core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java:
```diff
@@ -69,14 +74,14 @@
         this.ch = ch;
         this.cfg = cfg;
-        this.mappedPath = mappedPath;
+        this.pathParams = ImmutableMap.copyOf(pathParams);
```
This will most likely end up with no-op since all `PathMapping` implementations return `ImmutableMap` already.

trustin commented on core/src/test/java/com/linecorp/armeria/server/http/HttpServerTest.java:
```diff
@@ -154,19 +154,19 @@ protected void configure(ServerBuilder sb) throws Exception {
                 @Override
                 protected void doGet(ServiceRequestContext ctx, HttpRequest req, HttpResponseWriter res) {
-                    final long delayMillis = Long.parseLong(ctx.mappedPath().substring(1));
+                    final long delayMillis = Long.parseLong(ctx.pathParam("delay"));
```
Wondering if I should add `ctx.pathIntParam(...)` and others. Thoughts?

hyangtack replied:
I think it's not necessary for now. A user may use the current `DynamicHttpService` for this kind of scenario, and may get the value as a parameter. So we can think about this again at the part 2.

trustin commented:
I eventually want to integrate `DynamicHttpService` into the core:

```java
ServerBuilder sb = ...;
sb.service(new MyAnnotatedService());
...

public class MyAnnotatedService {
    @Get
    @Path("/users/:userId")
    public CompletableFuture<UserInfo> getUserInfo(@PathParam("userId") String userId) { ... }
}
```

hyangtack commented:
LGTM 👍  I will review again after your next work.

trustin commented:
Thanks, @hyangtack. Any other comments, people?

anuraaga commented on core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java:
```diff
@@ -48,9 +52,39 @@
+     * @throws IllegalStateException if the {@link PathMapping} associated with this request does not have
+     *                               a prefix
+     */
+    default String pathWithoutPrefix() {
```
Maybe worth a non-volatile memo

trustin replied:
Like..?

anuraaga replied:
It would have to be on DefaultRequestContext but just the normal
```
private String pathWithoutPrefix;

String pathWithoutPrefix() {
  if (pathWithoutPrefix != null) {
    return pathWithoutPrefix;
  }
  pathWithoutPrefix = ...substring...
  return pathWithoutPrefix;
}
```

Not a big deal, but this method smells like it shouldn't have logic when calling it multiple times.

trustin replied:
Ah, memoization. OK.

anuraaga replied:
Sorry, yeah `memo` was too ambiguous ;)

trustin replied:
Added back `mappedPath()` and marked as deprecated. Its behavior is different from the old `mappedPath()` though.

hyangtack commented:
I found the current `DynamicHttpService` doesn't accept `@Path("/no-path-variable")` annotation because of the below link. I think this should be accepted in your new implementation. :)

https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java#L83
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/common/http/PathParamExtractor.java#L155

trustin commented:
> I found the current `DynamicHttpService` doesn't accept `@Path("/no-path-variable")` annotation because of the below links. I think this should be accepted in your new implementation. :)

Yeah, that's the plan.

anuraaga commented:
I realized losing mappedPath seems to remove the ability to do internal redirects with armeria. One example is this mapping that routes all urls to index.html for handling by a SPA

https://github.com/curioswitch/curiostack/blob/master/common/server/framework/src/main/java/org/curioswitch/common/server/framework/staticsite/StaticSiteService.java#L46

And another I was thinking of is repointing /legacy/rest/method to new.grpc.Service/Method for migrating APIs.

Perhaps it makes sense to allow mappings to continue to return a mapped path like current with the addition of the params map?

trustin commented:
> I realized losing mappedPath seems to remove the ability to do internal redirects with armeria.

Shouldn't this be done with a `302 Found` redirect response? Otherwise the `index.html` will be served from any location, causing an unexpected behavior. For example, a user can request `/foo/bar/baz/qux.html`, `qux.html` may refer to a script path `./qux.js` and `./qux.js` will be mapped to `/index.html` again.

trustin commented:
> And another I was thinking of is repointing /legacy/rest/method to new.grpc.Service/Method for migrating APIs.

This indeed sounds like a good reason to use internal redirects. I'm not sure if it is a good idea to implement an internal redirect with `PathMapping`, though. Would it be better providing internal redirects as a core functionality?

anuraaga commented:
That's why the static path does need to be specially mapped as well as it is in that service - this is normal for SPAs I think. Routing for SPAs happens within index.html - so all URLs for the service need to be handled by the same file - the URLs are correct so they can't be changed via the http redirect (which would cause the JS routing to not work since it would always see the same url, /), it should be an internal redirect.

There could be other implementations for this specific use case (single file static service), but internal redirects are common in server frameworks (nginx and apache call it url rewriting) so losing it seems like a regression.

anuraaga commented:
Having internal redirects as a separate feature would be ok too - I hope services could still be able to set them up agnostic to what prefix they are bound at, probably through a serviceAdded callback.

I agree it feels a little weird doing it in PathMapping, but after trying it, it was actually quite elegant how the redirects fit well into the path hierarchy, which is built on PathMapping.

trustin commented:
That makes sense. What I don't like about using `mappedPath` as a way to do an internal redirection is that `mappedPath` is meant for stripping the context path of a service from a request path, not for an internal redirection. How about just making `PathMapping` return a new struct that contains both path parameters and the path so that `PathMapping` can replace a request path completely?

anuraaga commented:
Yep, that's exactly the implementation I was thinking :)

trustin commented:
@anuraaga

- Added `PathMappingResult` and let `PathMapping` replace `path` and `query` completely for internal redirection. 
- Added RequestContext.query()
- RequestContext.path() does not include a query string anymore.
- Added RequestLog.query()

anuraaga commented on core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:
```diff
@@ -61,14 +63,15 @@
     public DefaultClientRequestContext(
             EventLoop eventLoop, SessionProtocol sessionProtocol, Endpoint endpoint,
-            String method, String path, String fragment, ClientOptions options, Object request) {
+            String method, String path, String query, @Nullable String fragment,
```
Think query is `@Nullable`

trustin replied:
Oops. Fixed.

trustin commented:
I also changed the type of `RequestContext.method()` from `String` to `HttpMethod`. It was always meant to be session-layer method name just in case it wasn't clear.

trustin commented:
`RequestContext.path()` and `RequestContext.query()` both now return a raw and undecoded string for consistency (and to give a user more control)

trustin commented:
@imasahiro Ready for a review.
@kojilin You may be interested as well.

hyangtack commented on core/src/main/java/com/linecorp/armeria/client/http/HttpRequestSubscriber.java:
```diff
@@ -152,8 +154,20 @@ private void writeFirstHeader() {
         }
 
+        final String pathWithQuery = firstNonNull(firstHeaders.path(), "/");
+        final String path;
```
I found `HttpServerHandler.validateAndSplitPath` also has a similar codes. :)

trustin replied:
Moved the common logic into `ArmeriaHttpUtil.splitPathAndQuery()`. Thanks!

trustin commented on core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java:
```diff
@@ -116,6 +119,11 @@ public final String path() {
     }
 
     @Override
+    public final String query() {
```
Is it necessary? Its super method is annotated with `@Nullable` already.

hyangtack replied:
Forget it. My mistake. :)

trustin commented on core/src/main/java/com/linecorp/armeria/server/PathMappingResult.java:
```diff
@@ -0,0 +1,121 @@
+    /**
+     * Returns the query mapped by the {@link PathMapping}.
+     */
+    public String query() {
```
Fixed.

hyangtack commented:
Thanks, @trustin! 😄

trustin commented:
Thanks!
